proposals/0000-hasfield-redesign.rst

```diff
@@ -0,0 +1,1193 @@
+``modifyPartialField`` would return the value unchanged.  The ``FieldTotal``
+class would allow an optics library to determine whether a particular field was
+total and hence whether it should produce a lens or an affine traversal.
+
+For now we propose not to include support for partial fields through classes
+like this, in the interests of minimizing complexity.
+
+
+Setting vs modification
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous design supported only ``setField :: a -> r -> r`` and not
+``modifyField :: (a -> a) -> r -> r``.  The latter generalises ``setField`` to
+allow modifying any ``a`` values in the datatype (of which there may be none).
+
+It is easy to implement ``setField`` in terms of ``modifyField``, but not vice
+versa, because we would need to define: ::
+
+  modifyFieldAlt :: forall x r a . (HasField x r a, SetField x r a) => (a -> a) -> r -> r
+  modifyFieldAlt f r = setField @x (f (getField @x r)) r
+
+This imposes an additional ``HasField`` constraint, and will necessarily be
+partial if ``getField`` is partial (whereas ``modifyField`` can in principle be
+total, although this will not be the case for automatically solved constraints,
+as discussed above).
+
+Thus we propose to use ``modifyField`` in the class, and define ``setField`` in
+terms of it.
+
+A consequence of this is that it is not possible to use ``SetField`` for types
+that are "write-only", e.g. where they do not contain a value for the field at
+all, and hence ``modifyField`` cannot be defined.
+
+
+Kind of field labels
+~~~~~~~~~~~~~~~~~~~~
+
+When ``HasField`` was originally introduced in `proposals #6
+<https://github.com/ghc-proposals/ghc-proposals/pull/6>`_, the kind of the
+parameter ``x`` representing the field label was polymorphic::
+
+  class HasField (x :: k) r a | x r -> a where ...
+
+While the class allows ``k :: Type`` to vary freely, ``HasField`` constraints
+will be solved only if it is instantiated to ``Symbol``.  Moreover,
+``OverloadedRecordDot`` and ``OverloadedRecordUpdate`` will only ever generate
+constraints using ``Symbol``.  Other possibilities were permitted in order to
+support hypothetical anonymous records libraries, which might support different
+kinds of fields, e.g. drawn from explicitly-defined enumerations.
+
+In principle it would be possible to simplify the class by specialising it to
+use ``Symbol`` rather than ``k``.  However we propose to retain the poly-kinded
+definition in the interests of generality and compatibility.  For example,
+the ``record-hasfield`` library makes use of the possibility to define label
+kinds other than ``Symbol``, allowing tuples of labels to be used for
+composition of fields.  In particular, it defines an instance like::
+
+  instance (HasField x1 r1 r2, HasField x2 r2 a2)
+      => HasField '(x1, x2) r1 a2
+
+which means ``getField @("foo", "bar")`` will be treated like the composition
+``getField @"bar" . getField @"foo"``.
```

Comment:
Another application I can imagine is to have instances for tuples

`instance HasField 1 (x,y)`

Not sure that it's particularly useful (unless we are able to use `u.1` as a notation), but putting it out there.

Reply:
Nice idea. In fact I imagine one could use generics to implement an instance of `HasField (n :: Nat) t` for tuples and/or record types...